Source/Core/Core/Debugger/HLEFormatter.cpp

```diff
@@ -0,0 +1,72 @@
+// Copyright 2021 Dolphin Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "HLEFormatter.h"
+
+#include <fmt/format.h>
+
+#include "Common/Logging/Log.h"
+#include "Core/HW/ProcessorInterface.h"
+#include "Core/PowerPC/PowerPC.h"
+
+namespace Core::Debug
+{
+namespace
+{
+using HLENamedArgs = fmt::dynamic_format_arg_store<fmt::format_context>;
+
+HLENamedArgs GetHLENamedArgs()
+{
+  auto args = HLENamedArgs();
+  std::string name;
+
+  for (int i = 0; i < 32; i++)
+  {
+    // General purpose registers (int)
+    name = "r" + std::to_string(i);
+    args.push_back(fmt::arg(name.c_str(), GPR(i)));
+
+    // Floating point registers (double)
+    name = "f" + std::to_string(i);
+    args.push_back(fmt::arg(name.c_str(), rPS(i).PS0AsU64()));
+    name += "_2";
+    args.push_back(fmt::arg(name.c_str(), rPS(i).PS1AsU64()));
+  }
+
+  // Special registers
+  args.push_back(fmt::arg("TB", PowerPC::ReadFullTimeBaseValue()));
+  args.push_back(fmt::arg("PC", PowerPC::ppcState.pc));
+  args.push_back(fmt::arg("LR", PowerPC::ppcState.spr[SPR_LR]));
+  args.push_back(fmt::arg("CTR", PowerPC::ppcState.spr[SPR_CTR]));
+  args.push_back(fmt::arg("CR", PowerPC::ppcState.cr.Get()));
+  args.push_back(fmt::arg("XER", PowerPC::GetXER().Hex));
+  args.push_back(fmt::arg("FPSCR", PowerPC::ppcState.fpscr.Hex));
+  args.push_back(fmt::arg("MSR", PowerPC::ppcState.msr.Hex));
+  args.push_back(fmt::arg("SRR0", PowerPC::ppcState.spr[SPR_SRR0]));
+  args.push_back(fmt::arg("SRR1", PowerPC::ppcState.spr[SPR_SRR1]));
+  args.push_back(fmt::arg("Exceptions", PowerPC::ppcState.Exceptions));
+  args.push_back(fmt::arg("IntMask", ProcessorInterface::GetMask()));
+  args.push_back(fmt::arg("IntCause", ProcessorInterface::GetCause()));
+  args.push_back(fmt::arg("DSISR", PowerPC::ppcState.spr[SPR_DSISR]));
+  args.push_back(fmt::arg("DAR", PowerPC::ppcState.spr[SPR_DAR]));
+  args.push_back(fmt::arg("Hash Mask", (PowerPC::ppcState.pagetable_hashmask << 6) |
+                                           PowerPC::ppcState.pagetable_base));
+
+  return args;
+}
+}
+
+std::string HLEFormatString(const std::string_view message)
+{
+  const HLENamedArgs args = GetHLENamedArgs();
+  try
+  {
+    return fmt::vformat(message, args);
+  }
+  catch (const fmt::format_error& e)
+  {
+    ERROR_LOG_FMT(MEMMAP, "Failed to format message: {}", e.what());
+    return message.data();
+  }
```

Comment:
Do they have a non-throwing version as well? Since we build without exceptions and all that...

Reply:
I haven't found one, yet. Based on the source: 
 - the exception is thrown using the macro `FMT_THROW` in the `throw_format_error` function,
 - that macro might call `FMT_ASSERT` instead of `throw` if exception is disabled (i.e. FMT_EXCEPTIONS is false),
 - if NDEBUG is set, `fmt::detail::ignore_unused` should noop.

I'll try to see if I can meet these requirements in the Windows build and get rid of the throwing version.

Reply:
I believe it's possible to redefine `FMT_THROW` and `FMT_ASSERT`, but that could be a bit fiddly.  It'd be nice to have them hooked in to Dolphin's panic alert system (in most cases) though.

Reply:
Sadly I don't we can because both `FMT_THROW` and `FMT_ASSERT` are used in `constexpr` functions _(in `fmt/core.h`)_ and our panic alert system isn't `constexpr`.